### PR TITLE
Handle different behavior of .array() in NumPy 2.0.0

### LINKED ISSUE #3723: NumPy 2.0 support
NumPy 2.0 is coming out soon ( https://github.com/numpy/numpy/issues/24300 ). NumPy 2.0.0rc1 packages for conda & wheels came out at the beginning of April ( https://github.com/numpy/numpy/issues/24300#issuecomment-2030603395 )

To prepare for NumPy 2.0, it might be worthwhile to start testing against NumPy 2 in CI

NumPy has put out [a migration guide]( https://numpy.org/devdocs/numpy_2_0_migration_guide.html#numpy-2-migration-guide ). More details are in [the release notes]( https://numpy.org/devdocs/release/2.0.0-notes.html )

If mlpack make use of NumPy's C API (and produces wheels that use it), having a release of mlpack with wheels (if provided) built against NumPy 2.0.0rc1 would be helpful to ensure NumPy 1 & 2 compatible wheels (as wheels built against NumPy 1 won't be compatible with NumPy 2). More details in [this NumPy 2 ABI doc]( https://numpy.org/devdocs/dev/depending_on_numpy.html#numpy-2-abi-handling )

Also as NumPy is tracking ecosystem support for NumPy 2.0, it would be helpful to share mlpack current support status in issue (with any plans): https://github.com/numpy/numpy/issues/26191

### PR DESCRIPTION
**Pull request by rcurtin**
This should fix #3723; the behavior of `np.array()` changed in NumPy 2.0.0 and now throws a `ValueError` if the `copy=False` argument is given and a copy cannot be avoided.  But we were using `copy=False` with the intention of "don't copy unless you need to"---where the more correct option is just to pass `None` instead.  According to the documentation that should be backwards-compatible with NumPy 1.x too; CI will tell us if that's true in just a few minutes. :)

@jakirkham if you happen to have a moment to review, it would be appreciated, but no worries if too much is going on. :+1:

### CODE REVIEW

**Review by zoq — APPROVED** [review]
Works on my end.

**Comment by rcurtin** [comment]
@mlpack-jenkins test this please

File: src/mlpack/bindings/python/mlpack/matrix_utils.py [hunk 1]
```diff
@@ -160,7 +160,10 @@ def to_matrix_with_info(x, dtype, copy=False):
       dims = len(x)
 
     d = np.zeros([dims])
-    out = np.array(x, dtype=dtype, copy=copy) # Try to avoid copy...
+    if copy:
+      out = np.asarray(x, dtype=dtype, copy=True)
+    else:
+      out = np.asarray(x, dtype=dtype, copy=None)
```
Comment (jakirkham) [inline src/mlpack/bindings/python/mlpack/matrix_utils.py:166]: Don't think `asarray` had the `copy` argument in NumPy 1. It isn't listed [in the NumPy 1 docs]( https://numpy.org/doc/1.26/reference/generated/numpy.asarray.html ). Also [NumPy 2 docs]( https://numpy.org/doc/2.0/reference/generated/numpy.asarray.html ) note `copy` is a new argument. Testing local with NumPy 1 am seeing an error. Am I missing something? Or are these meant to be `array` instead?

**Review by jakirkham — COMMENTED** [review]
Thanks Ryan! 🙏

Had a question below

File: src/mlpack/bindings/python/mlpack/matrix_utils.py (hunk 1 above)
Comment (rcurtin) [reply src/mlpack/bindings/python/mlpack/matrix_utils.py]: Ahh, thank you!  I did mean to write `array()` there, that was the documentation I checked to come up with this fix.  Just pushed a fix now.

Comment (jakirkham) [reply src/mlpack/bindings/python/mlpack/matrix_utils.py]: All good 😊 Thanks Ryan 🙏

File: src/mlpack/bindings/python/mlpack/matrix_utils.py [hunk 2]
```diff
@@ -160,7 +160,10 @@ def to_matrix_with_info(x, dtype, copy=False):
       dims = len(x)
 
     d = np.zeros([dims])
-    out = np.array(x, dtype=dtype, copy=copy) # Try to avoid copy...
+    if copy:
+      out = np.array(x, dtype=dtype, copy=True)
+    else:
+      out = np.array(x, dtype=dtype, copy=None)
```
Comment (jakirkham) [inline src/mlpack/bindings/python/mlpack/matrix_utils.py:166]: Think the `None` option is [new in NumPy 2]( https://numpy.org/doc/stable/release/2.0.0-notes.html#new-copy-keyword-meaning-for-array-and-asarray-constructors ). It [isn't mentioned in NumPy 1]( https://numpy.org/doc/1.26/reference/generated/numpy.array.html ). So likely there is another condition needed to check for NumPy 2 and use the `copy` option then and if not drop the `copy` argument. That is assuming that is the behavior intended here

Comment (rcurtin) [reply src/mlpack/bindings/python/mlpack/matrix_utils.py]: Thanks!  I thought I could get away with it (and was depending on CI to tell me if I was wrong, which it did), but you're right.  Should be fixed in 57d2e8.

**Review by shrit — APPROVED** [review]

**Comment by jakirkham** [comment]
Thanks all! 🙏

**Pull request merged**